var/spack/repos/builtin/packages/espresso/package.py

```diff
@@ -85,6 +88,16 @@ class Espresso(Package):
     conflicts('+elpa', when='@:5.4.0')
     conflicts('+hdf5', when='@:5.4.0')
 
+    # To build with QMCPACK's hdf5 interface we require QE 5.3.0
+    # and hdf5. 
+    conflicts('+qmcpackconv', when='@5.3.1:')
+    conflicts('+qmcpackconv', when='@:5.2.999')
+    depends_on('hdf5', when='+qmcpackconv')
+    depends_on('hdf5+mpi', when='+qmcpackconv+mpi')
+    depends_on('hdf5~mpi', when='+qmcpackconv~mpi')
```

Comment:
OT: What does the patch do? I see that at the moment we have no other choice here, but it seems to me quite a bad design idea if a package gains:

- 1 variant directive
- 1 patch directive
- 2 conflict directives
- 3 depends_on directives

just because a *dependent* needs to patch it for its own use. I think a better long term solution involves:

1. taking patches into account in the hashing mechanism
2. permit to apply patches to dependencies

At that point we might write in `qmcpack`:
```python
depends_on(
    'espresso@5.3.0~elpa+mpi', 
    when='+mpi',
    patch='<url>',
    patch_md5='ae395964e4d124f5eef48a53001513a0'
)
```
@adamjstewart @tgamblin Thoughts?

Reply:
@alalazo: I've wanted something like this for a while.  There are a fair number of packages that really want to manage their own patches on their dependencies, e.g.:

* `hpctoolkit`, if its `hpctoolkit-externals` dependency were ever to move entirely into Spack
* `openspeedshop` does similar stuff too

I also wish people would just upstream these types of changes, but sometimes that is not possible, and we don't know enough about this particular situation to say what should happen.  It's not really our job in Spack to know or say.

The only tricky bit is that this is really making a custom version of the dependency just for the dependent, so the patch should modify the dependency's hash.  Patches, in general, should probably modify the `Spec` hash.  We don't do that currently because patches are owned by package authors.

Syntax-wise, why not reuse the existing patch directive?

```python
depends_on(
    'espresso@5.3.0~elpa+mpi', 
    patch('<url>', md5='ae395964e4d124f5eef48a53001513a0'),
    when='+mpi')
```

That way we don't have to invent new prefixed `kwargs` or add them every time `patch` changes.  Could also allow this:

```python
depends_on(
    'espresso@5.3.0~elpa+mpi', 
    when='+mpi',
    patch=patch('<url>', md5='ae395964e4d124f5eef48a53001513a0'))
```

which doesn't require you to put `patch` first (in the regular args) but does feel more redundant.

Reply:
Note that patches for one project may break another project. I'd recommend making variants for these massive patches and I'd really like to see all patches annotated with either a "here's the patch we're backporting" or "here's where the patch has been submitted for merging" URL.

Reply:
@mathstuf 

> Note that patches for one project may break another project

I think we are on the same page here. What we were discussing above is that *if* `qmcpack` needs to patch `espresso` for its own purposes, then the patch should reside in the `qmcpack` package and the package itself should be able to apply it to the dependency.

> I'd really like to see all patches annotated with either a "here's the patch we're backporting" or "here's where the patch has been submitted for merging" URL.

That one for me is a :100: on the intentions, but - unfortunaley - we have also other use cases like "here's a patch found in a blog somewhere that works for me" or "here's a bunch of patches because [the maintainer likes to release that way](http://www.nwchem-sw.org/index.php/Download)"

@tgamblin 
> Could also allow this:
> ```python
> depends_on(
>    'espresso@5.3.0~elpa+mpi', 
>    when='+mpi',
>    patch=patch('<url>', md5='ae395964e4d124f5eef48a53001513a0'))
> ```
> which doesn't require you to put patch first (in the regular args) but does feel more redundant.

Why not just this then:
```python
depends_on(
    'espresso@5.3.0~elpa+mpi', 
    when='+mpi',
    patch=('<url>', md5='ae395964e4d124f5eef48a53001513a0')
)
```
? `patch=` will capture the arguments for the call to `patch` that will be executed later at some point.

Reply:
Your example's not legal Python :).  Can't have arg= in a tuple; only in a function call.

I have a PR in the works for patching dependencies and adding to the spec.

Reply:
Ahem, I meant a dictionary - like anybody would believe this...

Reply:
@tgamblin Trying to follow the discussions. Sounds like I am waiting for Todd's PR request. After that, I will remove that bit of code that is going into espresso/package.py and adding the proposed code into qmcpack/package.py. Overall, I think it does make more sense - its a QMCPACK (owner) patch of Quantum ESPRESSO (consumer). 

Also, I will mention this extra complication that is not on our critical path.

We do have patches for older versions of espresso, but it is unlikely that anyone would use them. If I want to pick a particular espresso version to patch, would that just be a variant?

Reply:
Yes, qmcpack would own its patches on espresso. I think this functionality would be useful for other packages as well.   Any reason you don't upstream your changes to espresso?

We should also be able to make the patches conditional on espresso's version.  I and @alalazo need to think about the syntax but I have details worked out for allowing packages to create custom patched dependency versions.

Reply:
@tgamblin @prckent That is a good question Todd. Quantum Espresso is an Open Source project from Europe. They do not use git and they do not have automated regression testing. There are some other challenges with respect to getting contributions back in Quantum Espresso, but they are best left for offline conversations. Having said that, we have not tried to get our patch in recently. There is a desire to do this better long term.

Reply:
Agreeing with @naromero77  - we would like to get these patches upstream in QE, as well as simple bugfixes to QE committed, but there are "challenges" as Nick describes. One of my projects did get a major performance improvement upstreamed this year, so we still have some hope.

Support for patching different versions of QE would be nice, but only the patch of the latest supported version (5.3) has real importance.

Reply:
> we have also other use cases like "here's a patch found in a blog somewhere that works for me" or "here's a bunch of patches because the maintainer likes to release that way"

Do we really want to support the former? Why are we supporting patches with no review, testing, or anything other than "it didn't break what *I* care about at least". For the latter, links to the upstream policy would be sufficient for me (e.g., Vim does the same thing). At least this seems to have sufficient testing from the `qmcpack` side (in the way that it seems upstream doesn't have much testing either).

As for private upstreams, a note saying "I've emailed the patch upstream" with a date (so that followup can be deferred if there hasn't been a response in a reasonable amount of time) would also be sufficient.

Reply:
On the QMCPACK side, we *do* have a strong test of the patch as one of our integration tests.

Reply:
@tgamblin At the risk of coming across as annoying, I want to make sure this potential modification to the patch and dependency function is still on your radar.

Reply:
@naromero77: It is! I've already started the PR.

Reply:
@tgamblin Thanks Todd. What is the PR#? (sorry, I tried looking for it and could not find it).